src/librustc_typeck/outlives/implicit_infer.rs

```diff
@@ -244,199 +262,34 @@ fn insert_required_predicates_to_be_wf<'tcx>(
-                        //    x:  &'a Vec<U>
-                        // }
-                        // ```
-                        //
-                        // Here `outlived_region = 'a` and `kind =
-                        // Vec<U>`.  Decomposing `Vec<U>` into
-                        // components would yield `U`, and we add the
-                        // where clause that `U: 'a`.
-                        let ty: Ty<'tcx> = tcx.mk_param(param_ty.idx, param_ty.name);
-                        required_predicates
-                            .insert(ty::OutlivesPredicate(ty.into(), outlived_region));
-                    }
-
-                    Component::Projection(proj_ty) => {
-                        // This would arise from something like:
-                        //
-                        // ```
-                        // struct Foo<'a, T: Iterator> {
-                        //    x:  &'a <T as Iterator>::Item
-                        // }
-                        // ```
-                        //
-                        // Here we want to add an explicit `where <T as Iterator>::Item: 'a`.
-                        let ty: Ty<'tcx> = tcx.mk_projection(proj_ty.item_def_id, proj_ty.substs);
-                        required_predicates
-                            .insert(ty::OutlivesPredicate(ty.into(), outlived_region));
-                    }
-
-                    Component::EscapingProjection(_) => {
-                        // As above, but the projection involves
-                        // late-bound regions.  Therefore, the WF
-                        // requirement is not checked in type definition
-                        // but at fn call site, so ignore it.
-                        //
-                        // ```
-                        // struct Foo<'a, T: Iterator> {
-                        //    x: for<'b> fn(<&'b T as Iterator>::Item)
-                        //              //  ^^^^^^^^^^^^^^^^^^^^^^^^^
-                        // }
-                        // ```
-                        //
-                        // Since `'b` is not in scope on `Foo`, can't
-                        // do anything here, ignore it.
-                    }
-
-                    Component::UnresolvedInferenceVariable(_) => bug!("not using infcx"),
+    debug!("def_id = {:?}", &def_id);
+    debug!("substs = {:?}", &substs);
+    debug!("explicit_map =  {:?}", explicit_map);
+    debug!("required_predicates = {:?}", required_predicates);
+    if let Some(explicit_predicates) = explicit_map.get(def_id) {
+        for outlives_predicate in explicit_predicates.iter() {
+            debug!("outlives_predicate = {:?}", &outlives_predicate);
+            let predicate = outlives_predicate.subst(tcx, substs);
+            debug!("predicate = {:?}", &predicate);
+
+            // FIXME explain why its ok to ignore Self.
+            if let (Some(self_ty), UnpackedKind::Type(ty)) = (ignore_self_ty, predicate.0.unpack())
```

Comment:
@nikomatsakis I need some help with explaining why its ok to ignore `Self` here.

---

Even though I am passing in Self and I was thinking more on our conversation about the different ways we could infer for TyDynamic:
- pass Self and dont infer
- pass self, infer and then remove Self predicates
- pass a Fake ty and dont worry about inferring

The 'Fake ty' seems like the cleanest choice since it will never be in `explicit_map`. By using `Self` or any other type, we need to mentally check and comment why its ok to not infer.